### PR TITLE
Add enum support to `rust` and skip none option serialization in clients

### PR DESCRIPTION
**Pull request by gferon**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh`, `./bin/security/{LANG}-petstore.sh` and `./bin/openapi3/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`~~, `3.4.x`, `4.0.x`~~. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

While generating rust clients, I got blocked by two issues, so here are the fixes.

The first commit adds a `serde` attribute to skip serialization if an optional field is unset - which means the field will be omitted instead of having `"option": null` in the request. This is desirable because on the other-side, null is not a valid value.

The second commit adds support for `enum` generation (the facilities to format name and values was already there in `RustClientCodegen.java) with a change to casing to be more rust-y.

### CODE REVIEW

**Comment by gferon** [comment]
Looks like I can't add reviewers myself, so tagging @frol, @farcaller or @bjgill to get a review.

File: modules/openapi-generator/src/main/resources/rust/model.mustache [hunk 1]
```diff
@@ -31,16 +43,7 @@ impl {{{classname}}} {
...
-// List of {{{name}}}
-//const (
-//    {{#allowableValues}}
-//    {{#enumVars}}
-//    {{{name}}} {{{classname}}} = "{{{value}}}"
-//    {{/enumVars}}
-//    {{/allowableValues}}
-//)
 {{/isEnum}}
+
```
Comment (bjgill) [inline modules/openapi-generator/src/main/resources/rust/model.mustache]: Is this extra line intentional?

File: modules/openapi-generator/src/main/resources/rust/model.mustache [hunk 2]
```diff
@@ -8,13 +8,25 @@
 #[allow(unused_imports)]
 use serde_json::Value;
 
+{{#isEnum}}
+#[derive(Debug, Serialize, Deserialize)]
+pub enum {{classname}} {
+{{#allowableValues}}
+{{#enumVars}}
+    #[serde(rename = "{{{value}}}")]
+    {{name}},
```
Comment (bjgill) [inline modules/openapi-generator/src/main/resources/rust/model.mustache]: Do we want to add `{{{description}}}` here as we do for structs below?

**Comment by gferon** [comment]
Thanks for the comments @bjgill, I just made fixes for both. I think we'll start using rust generated clients a bit more, so you'll probably hear again from me!

**Comment by wing328** [comment]
@gferon can you please run `./bin/rust-petstore.sh` (or the corresponding batch file under `bin\windows`) to update the Rust petstore samples?

**Comment by gferon** [comment]
@wing328 done, though I'm now trying to understand why the `enum` won't work with `petstore.yaml` - is it possible that it's because it's a OpenAPI 2.0 spec?

**Comment by wing328** [comment]
@gferon are you free for a quick chat (IM) to discuss the issue via https://gitter.im (ID: wing328) or other channels today?

**Comment by wing328** [comment]
> is it possible that it's because it's a OpenAPI 2.0 spec?

Both OpenAPI 2.0 and 3.0 spec support enum.

Can you elaborate on the issue?

**Comment by gferon** [comment]
@wing328 I'm very sorry about the time it took me to get back to this one. I believe the latest stage is the one I wanted to achieve :) feel free to review again.

File: modules/openapi-generator/src/main/resources/rust/README.mustache [hunk 3]
```diff
@@ -40,69 +40,12 @@ Class | Method | HTTP request | Description
...
 
 ```
-    import "golang.org/x/oauth2"
-
-    / .. Perform OAuth2 round trip request and obtain a token .. //
-
-    tokenSource := oauth2cfg.TokenSource(createContext(httpClient), &token)
-    auth := context.WithValue(oauth2.NoContext, sw.ContextOAuth2, tokenSource)
-    r, err := client.Service.Operation(auth, args)
+cargo doc --open
```
Comment (bjgill) [inline modules/openapi-generator/src/main/resources/rust/README.mustache]: ❤️  the docs updates!

**Comment by bjgill** [comment]
I'm just checking that the samples are up to date, and then I'll merge.

**Comment by wing328** [comment]
@bjgill thanks for reviewing and merging the PR 👍

**Comment by wing328** [comment]
Thanks again for the PR, which has been included in the v4.0.2 release: https://twitter.com/oas_generator/status/1141610197766426626

**Pull request merged**